File: apps/desktop/src/lib/memory.ts
import { invoke } from "@tauri-apps/api/core";

export type MemoryTraceType = "working" | "episodic" | "semantic";

export type MemoryConsolidationState =
  | "none"
  | "candidate"
  | "approved"
  | "rejected"
  | "archived";

export type MemoryGraphNode = {
  id: string;
  kind: string;
  title: string;
  body: string | null;
  traceType: MemoryTraceType;
  consolidationState: MemoryConsolidationState;
};

export type MemoryGraphEdge = {
  id: string;
  sourceId: string;
  targetId: string;
  relation: string;
};

export type MemoryGraph = {
  nodes: MemoryGraphNode[];
  edges: MemoryGraphEdge[];
};

export type MemoryScope = {
  id: string;
  title: string;
  kind: string;
  workflowId: string | null;
  sessionId: string | null;
  agentId: string | null;
};

export type MemoryReviewSurface = "chat" | "workflow";

export type MemoryReviewDecision =
  | "promote_semantic"
  | "keep_episodic"
  | "reject";

export type MemoryCandidate = {
  id: string;
  nodeId: string;
  title: string;
  surface: MemoryReviewSurface;
  ownerId: string;
  suggestedSchemaId: string | null;
  confidence: number;
  reason: string;
  evidenceCount: number;
};

export async function loadMemoryGraph(scopeId?: string | null): Promise<MemoryGraph> {
  return invoke<MemoryGraph>(
    "load_memory_graph",
    scopeId ? { scopeId } : {},
  );
}

export async function listMemoryScopes(): Promise<MemoryScope[]> {
  return invoke<MemoryScope[]>("list_memory_scopes");
}

export async function updateMemoryNode(
  nodeId: string,
  title: string,
  body: string | null,
): Promise<MemoryGraphNode> {
  return invoke<MemoryGraphNode>("update_memory_node", { nodeId, title, body });
}

export async function deleteMemoryNode(nodeId: string): Promise<void> {
  return invoke("delete_memory_node", { nodeId });
}

export async function createMemoryEdge(
  edgeId: string,
  sourceId: string,
  targetId: string,
  relation: string,
): Promise<MemoryGraphEdge> {
  return invoke<MemoryGraphEdge>("create_memory_edge", {
    edgeId,
    sourceId,
    targetId,
    relation,
  });
}

export async function deleteMemoryEdge(edgeId: string): Promise<void> {
  return invoke("delete_memory_edge", { edgeId });
}

export async function listPendingMemoryCandidates(
  surface: MemoryReviewSurface,
  ownerId: string,
): Promise<MemoryCandidate[]> {
  return invoke<MemoryCandidate[]>("list_pending_memory_candidates", {
    surface,
    ownerId,
  });
}

export async function reviewMemoryCandidate(
  candidateId: string,
  decision: MemoryReviewDecision,
): Promise<void> {
  return invoke("review_memory_candidate", { candidateId, decision });
}

File: crates/nuka-domain/src/memory.rs
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct MemoryGraph {
    pub nodes: Vec<MemoryGraphNode>,
    pub edges: Vec<MemoryGraphEdge>,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum MemoryTraceType {
    Working,
    Episodic,
    #[default]
    Semantic,
}

impl MemoryTraceType {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Working => "working",
            Self::Episodic => "episodic",
            Self::Semantic => "semantic",
        }
    }

    pub fn from_str(value: &str) -> Result<Self, String> {
        match value {
            "working" => Ok(Self::Working),
            "episodic" => Ok(Self::Episodic),
            "semantic" => Ok(Self::Semantic),
            _ => Err(format!("unknown memory trace type: {value}")),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum MemoryConsolidationState {
    #[default]
    None,
    Candidate,
    Approved,
    Rejected,
    Archived,
}

impl MemoryConsolidationState {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::None => "none",
            Self::Candidate => "candidate",
            Self::Approved => "approved",
            Self::Rejected => "rejected",
            Self::Archived => "archived",
        }
    }

    pub fn from_str(value: &str) -> Result<Self, String> {
        match value {
            "none" => Ok(Self::None),
            "candidate" => Ok(Self::Candidate),
            "approved" => Ok(Self::Approved),
            "rejected" => Ok(Self::Rejected),
            "archived" => Ok(Self::Archived),
            _ => Err(format!("unknown memory consolidation state: {value}")),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MemorySurface {
    Chat,
    Workflow,
}

impl MemorySurface {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Chat => "chat",
            Self::Workflow => "workflow",
        }
    }

    pub fn from_str(value: &str) -> Result<Self, String> {
        match value {
            "chat" => Ok(Self::Chat),
            "workflow" => Ok(Self::Workflow),
            _ => Err(format!("unknown memory surface: {value}")),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ReviewDecision {
    PromoteSemantic,
    KeepEpisodic,
    Reject,
}

impl ReviewDecision {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::PromoteSemantic => "promote_semantic",
            Self::KeepEpisodic => "keep_episodic",
            Self::Reject => "reject",
        }
    }

    pub fn from_str(value: &str) -> Result<Self, String> {
        match value {
            "promote_semantic" => Ok(Self::PromoteSemantic),
            "keep_episodic" => Ok(Self::KeepEpisodic),
            "reject" => Ok(Self::Reject),
            _ => Err(format!("unknown review decision: {value}")),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MemoryNodeKind {
    Workflow,
    Session,
    Agent,
    Message,
    Fact,
}

impl MemoryNodeKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Workflow => "workflow",
            Self::Session => "session",
            Self::Agent => "agent",
            Self::Message => "message",
            Self::Fact => "fact",
        }
    }

    pub fn from_str(value: &str) -> Result<Self, String> {
        match value {
            "workflow" => Ok(Self::Workflow),
            "session" => Ok(Self::Session),
            "agent" => Ok(Self::Agent),
            "message" => Ok(Self::Message),
            "fact" => Ok(Self::Fact),
            _ => Err(format!("unknown memory node kind: {value}")),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryGraphNode {
    pub id: String,
    pub kind: MemoryNodeKind,
    pub title: String,
    pub body: Option<String>,
    pub trace_type: MemoryTraceType,
    pub consolidation_state: MemoryConsolidationState,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryGraphEdge {
    pub id: String,
    pub source_id: String,
    pub target_id: String,
    pub relation: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryScope {
    pub id: String,
    pub name: String,
    pub workflow_id: Option<String>,
    pub session_id: Option<String>,
    pub agent_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryNodeSummary {
    pub id: String,
    pub kind: MemoryNodeKind,
    pub title: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryNodeDetail {
    pub id: String,
    pub kind: MemoryNodeKind,
    pub title: String,
    pub body: Option<String>,
    pub trace_type: MemoryTraceType,
    pub consolidation_state: MemoryConsolidationState,
    pub related_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MemoryCandidate {
    pub id: String,
    pub node_id: String,
    pub title: String,
    pub surface: MemorySurface,
    pub owner_id: String,
    pub suggested_schema_id: Option<String>,
    pub confidence: f32,
    pub reason: String,
    pub evidence_count: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryCandidateEvidence {
    pub id: String,
    pub candidate_id: String,
    pub detail: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemorySnapshot {
    pub id: String,
    pub node_id: String,
    pub title: String,
    pub body: Option<String>,
    pub trace_type: MemoryTraceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MemoryReviewAction {
    pub id: String,
    pub candidate_id: String,
    pub node_id: String,
    pub decision: ReviewDecision,
}

#[cfg(test)]
mod tests {
    #[test]
    fn memory_node_kind_rejects_unknown_values() {
        let error = super::MemoryNodeKind::from_str("timeline").unwrap_err();

        assert_eq!(error.to_string(), "unknown memory node kind: timeline");
    }

    #[test]
    fn memory_trace_type_rejects_unknown_values() {
        let error = super::MemoryTraceType::from_str("flashbulb").unwrap_err();

        assert_eq!(error.to_string(), "unknown memory trace type: flashbulb");
    }
}
